src/exp.ml

```diff
@@ -904,20 +957,23 @@ and of_let
       | _ -> true
       end ->
       Type.of_typ_expr true typ_vars exp_type >>= fun (_, _, new_typ_vars) ->
-      return (Name.Set.cardinal new_typ_vars <> 0)
+      return (List.length new_typ_vars <> 0)
     | _ -> return true
     end >>= fun is_function ->
     begin match cases with
     | [{ vb_pat = p; vb_expr = e1; vb_attributes; _ }] when not is_function ->
       let* attributes = Attribute.of_attributes vb_attributes in
-      Pattern.of_pattern p >>= fun p ->
-      of_expression typ_vars e1 >>= fun e1 ->
+      let* p_typ = Type.of_type_expr_without_free_vars (p.pat_type) in
+      let* p = Pattern.of_pattern p in
+      let* e1_typ = Type.of_type_expr_without_free_vars (e1.exp_type) in
+      let* e1 = of_expression typ_vars e1 in
+      let dep_match = { cast = p_typ; args = []; motive = e1_typ } in
       begin match p with
       | Some (Pattern.Variable x) -> return (LetVar (None, x, [], e1, e2))
       | Some p ->
         let is_with_default_case = Attribute.has_match_with_default attributes in
-        return (Match (e1, [p, None, e2], is_with_default_case))
-      | None -> return (Match (e1, [], false))
+        return (Match (e1, Some dep_match, [p, None, e2], is_with_default_case))
```

Comment:
Should dependent match be active only with the GADT tag attribute? Or removed as a first approximation, expecting the user to write a explicit `match` with a single case and a GADT tag attribute?